pkg/config/utils/endpoints.go

```diff
@@ -90,29 +88,98 @@ func GetMainEndpointBackwardCompatible(c pkgconfigmodel.Reader, prefix string, d
 	return prefix + pkgconfigsetup.DefaultSite
 }
 
+// MakeEndpoints takes a map of domain to apikeys and a config path root and converts this to
+// a map of domain to Endpoint structs.
+func MakeEndpoints(endpoints map[string][]string, root string) map[string][]Endpoint {
+	result := map[string][]Endpoint{}
+	for url, keys := range endpoints {
+		// Remove any empty API keys.
+		// We don't need to hold on to an endpoint with an empty API key to track if a
+		// secret has been updated since secrets can never be empty in the first place.
+		trimmed := []string{}
+		for _, key := range keys {
+			trimmedAPIKey := strings.TrimSpace(key)
+			if trimmedAPIKey != "" {
+				trimmed = append(trimmed, trimmedAPIKey)
+			}
+		}
+
+		if len(trimmed) > 0 {
+			result[url] = []Endpoint{{
+				ConfigSettingPath: root,
+				APIKeys:           trimmed,
+			}}
+		}
+	}
+
+	return result
+}
+
+// DedupEndpoints takes a map of domain to Endpoints and dedupes it so that we
+// only have a map of domain to (deduped) api keys.
+// Endpoints with no api keys are removed.
+// This needs to be a separate process to loading because we need to keep track
+// of the endpoints with the API config location to know when they have been
+// refreshed.
+func DedupEndpoints(endpointMap map[string][]Endpoint) map[string][]string {
+	keysPerDomain := map[string][]string{}
+
+	for domain, endpoints := range endpointMap {
+		dedupedAPIKeys := DedupEndpoint(endpoints)
+		if len(dedupedAPIKeys) > 0 {
+			keysPerDomain[domain] = dedupedAPIKeys
+		} else {
+			log.Infof("No API key provided for domain \"%s\", removing domain from endpoints", domain)
+		}
+	}
+
+	return keysPerDomain
+}
+
+// DedupEndpoint takes a single array of endpoints and returns an array of unique
+// api keys that they contain.
+func DedupEndpoint(endpoints []Endpoint) []string {
+	dedupedAPIKeys := make([]string, 0)
+	seen := make(map[string]bool)
+	for _, endpoint := range endpoints {
+		for _, apiKey := range endpoint.APIKeys {
+			if _, ok := seen[apiKey]; !ok {
+				seen[apiKey] = true
+				dedupedAPIKeys = append(dedupedAPIKeys, apiKey)
+			}
+		}
+	}
+
+	return dedupedAPIKeys
+}
+
 // GetMultipleEndpoints returns the api keys per domain specified in the main agent config
-func GetMultipleEndpoints(c pkgconfigmodel.Reader) (map[string][]string, error) {
+func GetMultipleEndpoints(c pkgconfigmodel.Reader) (map[string][]Endpoint, error) {
 	ddURL := GetInfraEndpoint(c)
 	// Validating domain
 	if _, err := url.Parse(ddURL); err != nil {
 		return nil, fmt.Errorf("could not parse main endpoint: %s", err)
 	}
 
-	keysPerDomain := map[string][]string{
-		ddURL: {
-			c.GetString("api_key"),
-		},
+	keysPerDomain := map[string][]Endpoint{
+		ddURL: {{
+			ConfigSettingPath: "api_key",
```

Comment:
Would it make sense to use `NewEndpoint` here and also at line 179?

Reply:
I hesitated to do so because the function parameters for `NewEndpoint` is just a bunch of strings. That function was mainly useful for tests. Creating the struct directly here allows us to use the struct field names as documentation.